File: app/graphql/mutations/create_product.rb
module Mutations
  class CreateProduct < BaseMutation
    argument :name, String, required: true
    argument :description, String, required: false
    argument :price, Float, required: true

    field :product, Types::ProductType, null: true
    field :errors, [String], null: true

    def resolve(name:, description: "", price:)
      product = Product.new(name: name, description: description, price: price)

      if product.save
        {
          product: product,
          errors:  [],
        }
      else
        {
          product: nil,
          errors:  product.errors.full_messages,
        }
      end
    end
  end
end

File: app/javascript/components/mutations/createProduct.js
import {gql} from 'apollo-boost';

export default gql`
  mutation createProduct($name: String!, $description: String, $price: Float!) {
    createProduct(name: $name, description: $description, price: $price) {
      product {
        id
        name
        description
        price
      }
    }
  }
`;
